# PAM's Descriptive Analysis

This file contains all the code & analysis developed for the paper titled XXX.

In this section we will analyse the PAM's datasets. 

In [90]:
# Packages
import os 
import io
import requests
import pandas as pd
import numpy as np
from langdetect import detect
import yaml
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', -1) # print long strings

In [4]:
# Read yaml file 
with open('config_file.yaml', 'r') as f:
    config = yaml.load(f)

### 1.1. Proposals Dataset

In [77]:
# Read dataset of proposals
df = pd.read_csv(config['ROOT_PATH'] + '/data/' + 'proposals_all.csv')
len(df)

10860

In [6]:
# Remove first column
df.drop(df.columns[[0]], axis=1, inplace=True)

In [7]:
# Analyse column types
df.dtypes

proposal          int64 
origin            object
scope             object
district          object
category          object
subcategory       object
author            object
author_name       object
created_at        object
votes             int64 
comments          int64 
url               object
status            object
title_es          object
title_ca          object
description_es    object
description_ca    object
dtype: object

In [8]:
# Change some column types
df.astype({'proposal': 'object'}).dtypes

proposal          object
origin            object
scope             object
district          object
category          object
subcategory       object
author            object
author_name       object
created_at        object
votes             int64 
comments          int64 
url               object
status            object
title_es          object
title_ca          object
description_es    object
description_ca    object
dtype: object

### Analysis of votes, comments, districts and categories

In [9]:
# Summary of variables
summary_list = [df.describe()] + \
               [df.groupby([c])[df.columns[0]].count() \
                for c in df.columns if df[c].dtype == 'object']

for i in summary_list:
    print(i)
    print()

           proposal         votes      comments
count  10860.000000  10860.000000  10860.000000
mean   5513.313628   15.201842     1.675046    
std    3139.352666   43.845657     5.708035    
min    3.000000      0.000000      0.000000    
25%    2799.750000   1.000000      0.000000    
50%    5515.500000   5.000000      0.000000    
75%    8230.250000   15.000000     1.000000    
max    10946.000000  1720.000000   337.000000  

origin
citizen         662 
citizenship     2783
meeting         4498
official        1300
organization    1617
Name: proposal, dtype: int64

scope
city        4518
district    6342
Name: proposal, dtype: int64

district
Ciutat Vella                971
Eixample                    539
Gràcia                   142
Gràcia                     571
Horta - Guinardó         116
Horta - Guinardó           490
Les Corts                   585
None                        824
Nou Barris                  444
Sant Andreu                 516
Sant Martí             

Proposals usually obtain a low number of votes (75% of them obtained at most 15 votes), although, the analysis shows that some proposals were very popular (with at most 1,720 votes). The distribution of comments also exposes that most of them (75%) obtained 0 or 1 comments. However, some of them were very commented (max. is 337).

The district with more proposals was Ciutat Vella, with at least 971 proposals. On the other hand, the Sarrià SAnt Gervasi was the least participative. 

The most popular category was Bon viure, the least 'Justicia Global'.

### Analysis of status, authors, and origin

In [10]:
# What does 0 and 1 mean in the status column?
df.loc[df['status'] == '0'] # status == '0' is equal to status == 'rejected'

df['status'] = df['status'].map({'0': 'rejected', '1': 'accepted'}) #clean status

In [16]:
df['status'].value_counts()

accepted    1540
rejected    514 
Name: status, dtype: int64

In [21]:
df['author_name'].value_counts()

Ajuntament de Barcelona                                1305
None                                                   846 
suport decidim.HG i Les Corts                          496 
suport decidim.barcelona                               391 
Suport General - Decidim Barcelona                     306 
Suport Ciutat Vella                                    290 
GR_Raons Públiques                                    254 
suportdecidim.santandreu                               237 
Suport decidim Activitat Econòmica                    223 
suport.decidim1                                        202 
decidim sarrià-sant gervasi                           186 
eixampledelibera                                       179 
suport decidim.barcelona2                              163 
Dsoriano                                               126 
suport decidim.actes de ciutat                         112 
suport decidim.salut                                   93  
Solidaritat Catalana per la Independènc

In [50]:
df['origin'].value_counts()

meeting         4498
citizenship     2783
organization    1617
official        1300
citizen         662 
Name: origin, dtype: int64

1,540 proposals were accepted, while 514 were rejected.
The most proposative authors were the city hall (1,300 proposals), civil servants and decidim meetings. However, there exists a citizen who authored 126 proposals.

### Accepted vs. rejected and (author, origin, district, category)

In [45]:
# Authors with most accepted proposals
df[['author_name', 'status']].pivot_table(index='author_name', columns='status', aggfunc=len, fill_value=0).sort_values(by='accepted', ascending = False)



status                                                           accepted  \
author_name                                                                 
None                                                             653        
Ajuntament de Barcelona                                          244        
AAVV l'Òstia                                                  17         
procomuns.net                                                    16         
suport decidim.HG i Les Corts                                    15         
PAD Mòbil - Sant Martí                                     14         
Associacio de Veïns i Veïnes Coll-Vallcarca                13         
Treballadores i treballadors dels Districtes                     12         
AAVV Barceloneta                                                 12         
Salvador Pastor Blasco                                           12         
Solidaritat Catalana per la Independència                     11         
treballadors i treballadores de Barcelona Activa                 11         
consellsectorialdonesgenerelgtbi                                 9          
Coordinadora d'associacions de veïns de Nou Barris            9          
Consell Barri Hostafrancs                                        8          
consellbarrisantsbadal                                           8          
Fundació Comtal                                               7          
ECOM                                                             7          
Associació de Veïns i Veïnes Clot-Camp de l'Arpa        7          
consellsdistrictesantmarti                                       7          
Amics de la Bici                                                 7          
Associació de Veïns i Veïnes Joan Maragall Guinardó  7          
AVV Diagonal Mar                                                 7          
Consell de Barri Sant Antoni                                     6          
consellbarrilamarina                                             6          
Sarrià-Sant Gervasi                                           6          
Unió d'Associacions de Veïns del Poble Sec                 6          
consellbarripoblesec                                             5          
consellsectorialcultura                                          5          
Casal d'Infants                                                  5          
...                                                             ..          
Cecilia Illa Malvehy                                             0          
Waldorapaz                                                       0          
Carme Martin                                                     0          
Xino Xano Associació de Lleure                                0          
Xnet                                                             0          
Sideroploide                                                     0          
Serena                                                           0          
Pau B                                                            0          
Recull Navas                                                     0          
Pedro Tarraga                                                    0          
PepBarna                                                         0          
Freebolero                                                       0          
Pepi Del Caño Martín-Zamorano                              0          
Piti Cervelló                                                 0          
Rafael Granero                                                   0          
Enrique Pico                                                     0          
Raquel Amado Mezquita                                            0          
Raquel C.                                                        0          
Regalim                                                          0       

In [47]:
# Authors with most rejected proposals
df[['author_name', 'status']].pivot_table(index='author_name', columns='status', aggfunc=len, fill_value=0).sort_values(by='rejected', ascending = False)


status                                                     accepted  rejected
author_name                                                                  
None                                                       653       193     
Salvador Pastor Blasco                                     12        12      
Ajuntament de Barcelona                                    244       10      
suport decidim.HG i Les Corts                              15        9       
Associacio de Veïns i Veïnes Coll-Vallcarca          13        9       
Sarrià-Sant Gervasi                                     6         6       
Consell Barri Hostafrancs                                  8         5       
AAVV Barceloneta                                           12        5       
Solidaritat Catalana per la Independència               11        5       
consellsdistrictesantmarti                                 7         5       
gentgrandistrictesantmarti                                 3         4       
Associació de Veïns i Veïnes Clot-Camp de l'Arpa  7         4       
procomuns.net                                              16        4       
AAVV l'Òstia                                            17        4       
Unió d'Associacions de Veïns del Poble Sec           6         3       
AVV Vila Olímpica                                       4         3       
Amics de la Bici                                           7         3       
Universitat de Barcelona                                   0         3       
consellbarrilamarina                                       6         3       
Treballadores i treballadors dels Districtes               12        2       
Associació de Veïns Illa Robador                     0         2       
Natàlia opina                                           0         2       
Consell de Barri Les Corts                                 2         2       
J. Alberto                                                 0         2       
consellsectorialgentgransantsmontjuïc                   2         2       
Jordi Reynés Garcés                                  0         2       
treballadors i treballadores de Barcelona Activa           11        2       
consellbarripoblesec                                       5         2       
espernallac                                                1         2       
Toni Campos                                                0         2       
...                                                       ..        ..       
Gonzalo Zaragoza                                           1         0       
Jaume Artigues Vidal                                       1         0       
Gerard Gual Gasulla                                        1         0       
Gemma Riquelme                                             1         0       
GR_Raons Públiques                                      2         0       
GENERA                                                     1         0       
Fundació Pere Tarrés                                 1         0       
Fundació Comtal                                         7         0       
Froilà                                                  1         0       
JTudó                                                   1         0       
Jaume Layola                                               2         0       
José Manuel                                             1         0       
Jordi Príncep                                           1         0       
Josep Mompin Poblet                                        3         0       
Josep Maria                                                1         0       
Josep M. Guillumet Anés                                 2         0       
Josep M Armengou                                           1         0       
Jordi Salabert                                             1         0       
Jordi R.              

In [49]:
# Origin with most rejected proposals
df[['origin', 'status']].pivot_table(index='origin', columns='status', aggfunc=len, fill_value=0).sort_values(by='rejected', ascending = False)


status        accepted  rejected
origin                          
citizen       422       240     
meeting       652       192     
organization  223       72      
official      243       10

In [51]:
# District with most rejected proposals
df[['district', 'status']].pivot_table(index='district', columns='status', aggfunc=len, fill_value=0).sort_values(by='rejected', ascending = False)


status                    accepted  rejected
district                                    
None                      633       191     
Sant Martí             123       60      
Ciutat Vella              149       51      
Sarrià - Sant Gervasi  39        45      
Les Corts                 66        43      
Sants Montjuïc         81        39      
Gràcia                 108       34      
Sant Andreu               76        31      
Nou Barris                73        8       
Eixample                  82        6       
Horta - Guinardó       110       6

In [52]:
# District with most rejected proposals
df[['category', 'status']].pivot_table(index='category', columns='status', aggfunc=len, fill_value=0).sort_values(by='rejected', ascending = False)


status                     accepted  rejected
category                                     
Bon viure                  699       210     
Transició ecològica  500       206     
Economia plural            213       51      
Bon govern                 121       42      
Justícia global         7         5

### Bilingual analysis

In [56]:
# Read the original dataset (without translations)
url = 'https://raw.githubusercontent.com/elaragon/metadecidim/master/proposals.tsv'
s = requests.get(url).content
df_origin = pd.read_csv(io.StringIO(s.decode('utf-8')), sep = '\t')

In [99]:
# Create column with language of proposal
def detect_lang(x):
    try:
        return detect(x)
    except:
        return 'null'
    
    
df_origin['language'] = df_origin['summary'].apply(detect_lang)
df_origin['language'].value_counts()

ca      10014
es      713  
pt      41   
it      36   
null    19   
sv      12   
fr      9    
en      4    
ro      3    
de      2    
lv      1    
no      1    
sk      1    
tl      1    
id      1    
nl      1    
fi      1    
Name: language, dtype: int64

In [101]:
# Analyse different languages
df_origin[['summary', 'language']].loc[(df_origin['language'] != 'ca') & (df_origin['language'] != 'es')]
df_origin.loc[((df_origin['language'] != 'ca') & (df_origin['language'] != 'es') & (df_origin['language'] != 'null')), 'language'] = 'ca' # all different languages are indeed catalan

In [102]:
df_origin['language'].value_counts()

ca      10128
es      713  
null    19   
Name: language, dtype: int64

Most proposals are written in Catalan.

In [107]:
# Merge this new column with df
df_origin_aux = df_origin[['id', 'total_positive_comments', 'total_neutral_comments', 'total_negative_comments', 'rejected_message', 'language']]
df = pd.merge(df, df_origin_aux, left_on=  ['proposal'], right_on= ['id'], how = 'left')

In [108]:
df.head()

Unnamed: 0  proposal        origin     scope           district               category                   subcategory author                                 author_name created_at  votes  comments                                                                                                                       url    status                                                                               title_es                                                                              title_ca                                                                                                                                                                                                                                                                                                                                                                    description_es  \
0  0           3591      citizenship   district  Nou Barris         Economia plural        Un nou lideratge públic      5187   VILLARRASA                                  2/27/2016  16     0         https://decidim.barcelona.cat/proposals/reduccion-del-ibi-de-las-viviendas-de-torre-baro                                  rejected  Reducción del IBI de las viviendas de TORRE BARÓN                                    Reducció de l'IBI dels habitatges de TORRE BARÓ                                     Revisar el alto valor que se paga del Impuesto de Bienes Inmuebles en Torre Baró, sobre todo aquellas fincas construidas a partir del año 2001.Revisión de la categoría de calles, del valor del suelo y de la construcción con el Objeto que el valor catastral se reduzca Hasta el punto que se corresponda con la realidad de las fincas y super entorno.                  
1  1           2747      organization  district  Gràcia            Transició ecològica  Medi ambient i espai públic  6285   NaN                                         2/17/2016  8      0         https://decidim.barcelona.cat/proposals/park-guell-prevencio-d-incendis-i-polissa-de-responsabilitat-civil-pels-visitant  rejected  Park Güell, prevencion de incendios y póliza de responsabilidad civil por visitante  Park Güell, Prevenció d'incendis i polissa de Responsabilitat Civil pèls visitant  Como park público hay que tener la prevención de atención epsl incendios con bocas de incendio por los bomberos. Póliza por los usuarios de pago y servicio sanitario y de atención médica durante todas las visitas durante el año                                                                                                                                         
2  2           3158      citizenship   district  Horta - Guinardó  Bon viure              Educació i coneixement       512    Solidaritat Catalana per la Independència  2/19/2016  373    17        https://decidim.barcelona.cat/proposals/institut-al-barri-d-horta                                                         accepted  Instituto en el barrio de Horta.                                                       Institut al barri d'Horta.                                                            Inicio del procedimiento para la construcción de un Instituto de secundaria en Horta, tal como reclaman las asociaciones del barrio.                                                                                                                                                                                                                                              
3  3           8968      citizenship   city      NaN                Transició ecològica  Mobilitat sostenible          15512  Archie                                      4/6/2016   1      0         https://decidim.barcelona.cat/proposals/bus-facilitar-entrada-gent-gran                                                   rejected  Bus: facilitar entrada ancianos                                                        Bus: facilitar entrada gent gran                                            

In [ ]:
# Count min, max, avg, mean words per title_es, etc.